src/shared/util.js

```diff
@@ -1219,16 +1219,111 @@ function MessageHandler(sourceName, targetName, comObj) {
   this.targetName = targetName;
   this.comObj = comObj;
   this.callbackIndex = 1;
+  this.streamId = 1;
   this.postMessageTransfers = true;
-  var callbacksCapabilities = this.callbacksCapabilities = Object.create(null);
-  var ah = this.actionHandler = Object.create(null);
+  let streamSinks = this.streamSinks = Object.create(null);
+  let streamControllers = this.streamControllers = Object.create(null);
+  let callbacksCapabilities = this.callbacksCapabilities = Object.create(null);
+  let ah = this.actionHandler = Object.create(null);
 
   this._onComObjOnMessage = (event) => {
-    var data = event.data;
+    let data = event.data;
     if (data.targetName !== this.sourceName) {
       return;
     }
-    if (data.isReply) {
+    if (data.stream) {
+      let sourceName = this.sourceName;
+      let targetName = data.sourceName;
+      let streamId = data.streamId;
+      switch (data.stream) {
+        case 'start_complete':
+          if (data.success === true || data.success === 'true') {
+            streamControllers[data.streamId].startCall.resolve();
+          } else {
+            streamControllers[data.streamId].startCall.reject();
+          }
+          break;
+        case 'pull_complete':
+          if (data.success === true || data.success === 'true') {
+            streamControllers[data.streamId].pullCall.resolve();
+          } else {
+            streamControllers[data.streamId].pullCall.reject();
+          }
+          break;
+        case 'pull':
+          // pull increases the desiredSize property of sink, so when it
+          // changes from -ve to +ve set ready property as resolved promise.
```

Comment:
Unless one already knows what `ve` means, then this comment is unfortunately not very helpful. (This applies to another comment below as well.)